File: apps/mobile/composeApp/src/commonMain/kotlin/media/conduit/mobile/ContinueWatching.kt
package media.conduit.mobile

import kotlin.math.ceil
import kotlin.time.Clock
import kotlin.time.Duration.Companion.days
import kotlin.time.Instant
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import media.conduit.mobile.account.ProgressSummary
import media.conduit.mobile.account.VideoItem
import media.conduit.mobile.account.progressByRecency

internal enum class ContinueWatchingKind { InProgress, NewEpisode, NextUp, Scheduled, CaughtUp }

internal data class ContinueWatchingPresentation(
    val kind: ContinueWatchingKind,
    val video: VideoItem? = null,
    val label: String? = null,
)

internal fun groupContinueWatching(items: List<ProgressSummary>): List<ProgressSummary> {
    return coreValue(buildJsonObject {
        put("type", "groupContinueWatching")
        put("progress", ProtocolJson.encodeToJsonElement(items))
    }).jsonArray.mapNotNull { index -> items.getOrNull(index.jsonPrimitive.content.toInt()) }
}

internal fun progressTitleUiKey(progress: ProgressSummary): String =
    progress.canonicalTitleId ?: "${progress.mediaType}\u001f${progress.mediaId}"

internal fun progressEpisodeUiKey(progress: ProgressSummary): String {
    val episode = progress.canonicalEpisodeKey
        ?: if (progress.season != null || progress.episode != null) {
            "s${progress.season ?: 0}:e${progress.episode ?: 0}"
        } else {
            progress.videoId
        }
    return "${progressTitleUiKey(progress)}\u001f$episode"
}

internal fun progressHistoryForDisplay(items: List<ProgressSummary>): List<ProgressSummary> =
    progressByRecency(items).distinctBy(::progressEpisodeUiKey)

internal fun continueWatchingPresentation(
    progress: ProgressSummary,
    videos: List<VideoItem>,
    today: String = Clock.System.now().toString().take(10),
    now: Instant = Clock.System.now(),
    watchedVideoIds: Set<String> = emptySet(),
): ContinueWatchingPresentation {
    val result = coreValue(buildJsonObject {
        put("type", "continueWatching")
        put("progress", ProtocolJson.encodeToJsonElement(progress))
        put("videos", ProtocolJson.encodeToJsonElement(videos))
        put("today", today)
        put("nowMs", now.toEpochMilliseconds())
        put("watchedVideoIds", ProtocolJson.encodeToJsonElement(watchedVideoIds))
    }).jsonObject
    val video = result["videoIndex"]?.jsonPrimitive?.content?.toIntOrNull()?.let(videos::getOrNull)
    return when (result.getValue("kind").jsonPrimitive.content) {
        "in-progress" -> ContinueWatchingPresentation(ContinueWatchingKind.InProgress, video)
        "new-episode" -> ContinueWatchingPresentation(ContinueWatchingKind.NewEpisode, video)
        "next-up" -> ContinueWatchingPresentation(ContinueWatchingKind.NextUp, video)
        "scheduled" -> ContinueWatchingPresentation(
            ContinueWatchingKind.Scheduled,
            video,
            video?.released?.let { releaseDateLabel(it, today) },
        )
        else -> ContinueWatchingPresentation(ContinueWatchingKind.CaughtUp, video)
    }
}

internal fun continueWatchingBadgeLabel(
    progress: ProgressSummary,
    presentation: ContinueWatchingPresentation,
    metadataReady: Boolean,
): String {
    if (!metadataReady && progress.mediaType == "series" && progress.watched) return "Next Up"

    return when (presentation.kind) {
        ContinueWatchingKind.InProgress -> remainingTimeLabel(progress)
            ?: progressPercentLabel(progress)
            ?: if (progress.mediaType == "series") "Next Up" else if (progress.watched) "Watched" else "Resume"
        ContinueWatchingKind.NewEpisode -> "New Episode"
        ContinueWatchingKind.NextUp -> "Next Up"
        ContinueWatchingKind.Scheduled -> presentation.label ?: "Upcoming"
        ContinueWatchingKind.CaughtUp -> "Caught up"
    }
}

internal fun remainingTimeLabel(progress: ProgressSummary): String? {
    if (progress.watched || progress.durationMs <= 0 || progress.durationMs <= progress.positionMs) return null
    val minutes = ceil((progress.durationMs - progress.positionMs) / 60_000.0).toInt().coerceAtLeast(1)
    if (minutes < 60) return "$minutes min left"
    val hours = minutes / 60
    val remainder = minutes % 60
    return if (remainder == 0) "${hours}h left" else "${hours}h ${remainder}m left"
}

private fun progressPercentLabel(progress: ProgressSummary): String? {
    if (progress.watched || progress.positionMs <= 0L || progress.durationMs <= 0L) return null
    val percent = ((progress.positionMs.toDouble() / progress.durationMs) * 100).toInt().coerceIn(1, 99)
    return "$percent% watched"
}

internal fun releaseDateLabel(released: String, today: String): String {
    val day = releaseDay(released) ?: return released
    if (day == today) return "Today"
    if (day == nextIsoDay(today)) return "Tomorrow"
    val month = day.substring(5, 7).toIntOrNull()?.let { monthNames.getOrNull(it - 1) }
    val date = day.substring(8, 10).toIntOrNull()
    return if (month != null && date != null) "$month $date" else released
}

internal fun episodeReleaseDateLabel(released: String?): String? {
    val day = releaseDay(released) ?: return released?.takeIf(String::isNotBlank)
    val year = day.substring(0, 4)
    val month = day.substring(5, 7).toIntOrNull()?.let { monthNames.getOrNull(it - 1) }
    val date = day.substring(8, 10).toIntOrNull()
    return if (month != null && date != null) "$month $date, $year" else released
}

private fun compareEpisodes(a: VideoItem, b: VideoItem): Int =
    compareValues(a.season, b.season).takeIf { it != 0 }
        ?: compareValues(a.episode, b.episode).takeIf { it != 0 }
        ?: a.id.compareTo(b.id)

internal fun compareEpisodeCoordinates(a: VideoItem, b: VideoItem): Int =
    compareValues(a.season, b.season).takeIf { it != 0 }
        ?: compareValues(a.episode, b.episode)

private fun releaseDay(value: String?): String? =
    value?.take(10)?.takeIf { isoDayPattern.matches(it) }

private fun VideoItem.hasAired(today: String, now: Instant): Boolean {
    if (available == false) return false
    releaseInstant()?.let { return it <= now }
    if (available == true) return true
    val day = releaseDay(released) ?: return available != false
    return day <= today
}

private fun VideoItem.isUpcomingRelease(today: String, now: Instant): Boolean {
    if (!released.orEmpty().contains("T")) {
        return releaseDay(released)?.let {
            it > today || (available == false && it == today)
        } == true
    }
    return releaseInstant()?.let { it > now } == true
}

private fun VideoItem.isReleaseAlert(progress: ProgressSummary, now: Instant): Boolean {
    val releaseTimestamp = releaseInstant() ?: return false
    val watchedTimestamp = parseInstant(progress.updatedAt) ?: return false
    return releaseTimestamp > watchedTimestamp && now - releaseTimestamp < 60.days
}

private fun VideoItem.releaseInstant(): Instant? = released?.let(::parseInstant)

private fun parseInstant(value: String): Instant? {
    val trimmed = value.trim()
    return runCatching { Instant.parse(trimmed) }.getOrNull()
        ?: releaseDay(trimmed)?.let { day ->
            runCatching { Instant.parse("${day}T00:00:00Z") }.getOrNull()
        }
}

internal fun nextEpisodeAfter(
    progress: ProgressSummary?,
    videos: List<VideoItem>,
    watchedVideoIds: Set<String> = emptySet(),
    today: String = Clock.System.now().toString().take(10),
    now: Instant = Clock.System.now(),
    watchedProgress: List<ProgressSummary> = emptyList(),
): VideoItem? {
    if (progress == null) return null
    val regular = orderedPlayableEpisodes(videos, today, now)
    val anchor = videos.firstOrNull { progressMatchesVideo(progress, it) }
    val anchorSeason = anchor?.season ?: progress.season
    val anchorEpisode = anchor?.episode ?: progress.episode
    if (anchorSeason == null || anchorEpisode == null) return null
    return regular.firstOrNull {
        (
            it.season!! > anchorSeason ||
                (it.season == anchorSeason && it.episode!! > anchorEpisode)
            ) &&
            it.id !in watchedVideoIds &&
            watchedProgress.none { watched ->
                watched.watched && progressMatchesVideo(watched, it)
            }
    }
}

internal fun orderedContinueWatchingEpisodes(
    videos: List<VideoItem>,
    today: String = Clock.System.now().toString().take(10),
    now: Instant = Clock.System.now(),
): List<VideoItem> = videos
    .filter {
        (it.season ?: 0) > 0 &&
            it.episode != null &&
            (it.available != false || it.isUpcomingRelease(today, now))
    }
    .sortedWith(compareBy<VideoItem>({ it.season }, { it.episode }, { it.id }))

internal fun orderedPlayableEpisodes(
    videos: List<VideoItem>,
    today: String = Clock.System.now().toString().take(10),
    now: Instant = Clock.System.now(),
): List<VideoItem> = orderedContinueWatchingEpisodes(videos, today, now)
    .filter { it.hasAired(today, now) }

internal fun orderedEpisodePickerVideos(
    videos: List<VideoItem>,
): List<VideoItem> = orderedContinueWatchingEpisodes(videos)

private fun nextIsoDay(day: String): String {
    val year = day.substring(0, 4).toInt()
    val month = day.substring(5, 7).toInt()
    val date = day.substring(8, 10).toInt()
    val daysInMonth = when (month) {
        2 -> if (year % 400 == 0 || (year % 4 == 0 && year % 100 != 0)) 29 else 28
        4, 6, 9, 11 -> 30
        else -> 31
    }
    if (date < daysInMonth) return isoDay(year, month, date + 1)
    if (month < 12) return isoDay(year, month + 1, 1)
    return isoDay(year + 1, 1, 1)
}

private fun isoDay(year: Int, month: Int, day: Int): String =
    "${year.toString().padStart(4, '0')}-${month.toString().padStart(2, '0')}-${day.toString().padStart(2, '0')}"

private val isoDayPattern = Regex("\\d{4}-\\d{2}-\\d{2}")
private val monthNames = listOf("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")

File: apps/web/src/lib/continue-watching.ts
import type { WatchProgress } from "./api"
import type { Video } from "./core"
import { coreValue } from "./core"

export type ContinueWatchingState =
  | { kind: "in-progress"; video?: Video }
  | { kind: "new-episode"; video: Video }
  | { kind: "next-up"; video: Video }
  | { kind: "scheduled"; video: Video; label: string }
  | { kind: "caught-up"; video?: Video }

export function groupContinueWatching(items: WatchProgress[]): WatchProgress[] {
  return coreValue<number[]>({ type: "groupContinueWatching", progress: items })
    .map((index) => items[index]!)
}

export function continueWatchingState(
  progress: WatchProgress,
  videos: Video[],
  now = new Date(),
  watchedVideoIds: ReadonlySet<string> = new Set(),
): ContinueWatchingState {
  const decision = coreValue<{
    kind: "in-progress" | "new-episode" | "next-up" | "scheduled" | "caught-up"
    videoIndex?: number
  }>({
    type: "continueWatching",
    progress,
    videos,
    today: localDay(now),
    nowMs: now.getTime(),
    watchedVideoIds: [...watchedVideoIds],
  })
  const video = decision.videoIndex == null ? undefined : videos[decision.videoIndex]
  if (decision.kind === "scheduled" && video?.released) {
    return { kind: "scheduled", video, label: releaseDateLabel(video.released, now) }
  }
  if (decision.kind === "new-episode" && video) return { kind: "new-episode", video }
  if (decision.kind === "next-up" && video) return { kind: "next-up", video }
  if (decision.kind === "in-progress") return { kind: "in-progress", video }
  return { kind: "caught-up", video }
}

export function continueWatchingBadge(
  item: WatchProgress,
  state: ContinueWatchingState,
  metadataReady = true,
): string {
  if (!metadataReady && item.mediaType === "series" && item.watched) return "Next Up"
  if (state.kind === "new-episode") return "New Episode"
  if (state.kind === "next-up") return "Next Up"
  if (state.kind === "scheduled") return state.label
  if (state.kind === "caught-up") return "Caught up"
  return remainingTimeLabel(item) ?? progressPercentLabel(item) ??
    (item.mediaType === "series" ? "Next Up" : item.watched ? "Watched" : "Resume")
}

export function remainingTimeLabel(
  progress: Pick<WatchProgress, "positionMs" | "durationMs" | "watched">,
): string | undefined {
  if (progress.watched || progress.durationMs <= progress.positionMs || progress.durationMs <= 0) {
    return undefined
  }
  const minutes = Math.max(1, Math.ceil((progress.durationMs - progress.positionMs) / 60_000))
  if (minutes < 60) return `${minutes} min left`
  const hours = Math.floor(minutes / 60)
  const remainder = minutes % 60
  return remainder ? `${hours}h ${remainder}m left` : `${hours}h left`
}

export function releaseDateLabel(released: string, now = new Date()): string {
  const releaseDay = calendarDay(released)
  if (!releaseDay) return released
  const today = localDay(now)
  if (releaseDay === today) return "Today"
  const tomorrow = new Date(now.getFullYear(), now.getMonth(), now.getDate() + 1)
  if (releaseDay === localDay(tomorrow)) return "Tomorrow"
  const [year, month, day] = releaseDay.split("-").map(Number)
  return new Intl.DateTimeFormat(undefined, { month: "short", day: "numeric" }).format(
    new Date(year!, month! - 1, day!),
  )
}

function calendarDay(value: string | undefined): string | undefined {
  if (!value) return undefined
  const isoDay = /^\d{4}-\d{2}-\d{2}/.exec(value)?.[0]
  if (isoDay) return isoDay
  const date = new Date(value)
  return Number.isNaN(date.getTime()) ? undefined : localDay(date)
}

function localDay(date: Date): string {
  const year = date.getFullYear()
  const month = String(date.getMonth() + 1).padStart(2, "0")
  const day = String(date.getDate()).padStart(2, "0")
  return `${year}-${month}-${day}`
}

function progressPercentLabel(
  progress: Pick<WatchProgress, "positionMs" | "durationMs" | "watched">,
): string | undefined {
  if (progress.watched || progress.positionMs <= 0 || progress.durationMs <= 0) return undefined
  const percent = Math.min(99, Math.max(1, Math.floor((progress.positionMs / progress.durationMs) * 100)))
  return `${percent}% watched`
}
